Examples/Algorithms/Geant4/include/ActsExamples/Geant4/EventStoreRegistry.hpp

```diff
@@ -34,24 +37,43 @@ class EventStoreRegistry {
   struct State {
     /// The current event store
     WhiteBoard* store = nullptr;
+
+    /// Use a std::set here because it allows for fast insertion and ensures
+    /// uniqueness. Thus particle collisions are detected early.
+    using ParticleContainer =
+        std::set<ActsFatras::Particle, ActsExamples::detail::CompareParticleId>;
+
     /// Initial and final particle collections
-    SimParticleContainer::sequence_type particlesInitial;
-    SimParticleContainer::sequence_type particlesFinal;
+    ParticleContainer particlesInitial;
+    ParticleContainer particlesFinal;
+
     /// The hits in sensitive detectors
     SimHitContainer::sequence_type hits;
+
     /// Tracks recorded in material mapping
     std::unordered_map<size_t, Acts::RecordedMaterialTrack> materialTracks;
+
     /// Particle hit count (for hit indexing)
     std::unordered_map<SimBarcode, std::size_t> particleHitCount;
-    /// Track ID to Barcode mapping
-    std::unordered_map<unsigned int, SimBarcode> trackIdMapping;
-    /// Track ID to root Track ID mapping
-    std::unordered_map<unsigned int, unsigned int> trackIdRootId;
-    /// Track ID generation counter
-    std::unordered_map<unsigned int, unsigned int> trackIdGenerationCount;
+    /// Geant4 Track ID to Barcode mapping
+    std::unordered_map<G4int, SimBarcode> trackIdMapping;
+    /// Geant4 Track ID subparticle counter (for subparticle indexing)
+    std::unordered_map<G4int, std::size_t> trackIdSubparticleCount;
 
     /// Data handles to read particles from the whiteboard
     const ReadDataHandle<SimParticleContainer>* inputParticles{nullptr};
+
+    /// Count particle ID collisions
+    std::size_t particleIdCollisionsInitial = 0;
+    std::size_t particleIdCollisionsFinal = 0;
+
+    /// Store subparticle count for {primVertex, secVertex, part, gen}
+    /// This is done using a pseudo-barcode that contains all fields but not the
+    /// subparticle counter. This can be used as key in a map to store the
+    /// subparticle information
+    using BarcodeWithoutSubparticle =
+        Acts::MultiIndex<uint64_t, 16, 16, 16, 16>;
```

Comment:
shouldn't we use the same bits as in the original barcode?

Reply:
but they must sum up to 64... and as long there is enough space it should be ok

Reply:
makes sense - my concern was that we could overflow by converting from this special bar code to the other

Reply:
we do never convert them into each other, see below